source/mir/combinatorics/package.d

```diff
@@ -0,0 +1,1150 @@
+    alloc = custom Allocator
+
+Returns:
+    Forward range, which yields the permutations
+*/
+Permutations permutations(size_t n) @safe pure nothrow
+{
+    assert(n, "must have at least one item");
+    return Permutations(new uint[n-1], new uint[n]);
+}
+
+/// ditto
+IndexedRoR!(Permutations, Range) permutations(Range)(Range r) @safe pure nothrow
+if (isInputRange!Range && hasLength!Range)
+{
+    return permutations(r.length).indexedRoR(r);
+}
+
+///
+pure @safe nothrow unittest
+{
+    import std.algorithm.comparison : equal;
+    import std.range : iota;
+
+    auto expectedRes = [[0, 1, 2],
+         [1, 0, 2],
+         [2, 0, 1],
+         [0, 2, 1],
+         [1, 2, 0],
+         [2, 1, 0]];
+
+    auto r = iota(3);
+    auto rp = permutations(r.length).indexedRoR(r);
+    assert(rp.equal!equal(expectedRes));
+
+    // direct style
+    auto rp2 = iota(3).permutations;
+    assert(rp2.equal!equal(expectedRes));
+}
+
+pure @safe nothrow unittest
+{
+    // is copyable?
+    import std.algorithm: equal;
+    import std.range: iota, dropOne;
+    auto a = iota(2).permutations;
+    assert(a.front.equal([0, 1]));
+    assert(a.save.dropOne.front.equal([1, 0]));
+    assert(a.front.equal([0, 1]));
+}
+
+/// ditto
+Permutations makePermutations(Allocator)(auto ref Allocator alloc, size_t n)
+{
+    assert(n, "must have at least one item");
+    import std.experimental.allocator: makeArray;
+    auto state = alloc.makeArray!uint(n - 1);
+    auto indices = alloc.makeArray!uint(n);
+    return Permutations(state, indices);
+}
+
+/*
+Disposes a Permutations object. It destroys and then deallocates the
+Permutations object pointed to by a pointer.
+It is assumed the respective entities had been allocated with the same allocator.
+
+Params:
+    alloc = Custom allocator
+    perm = Permutations object
+*/
+void dispose(Allocator)(auto ref Allocator alloc, auto ref Permutations perm)
+{
+    import std.experimental.allocator: dispose;
+    dispose(alloc, perm.state);
+    dispose(alloc, perm.indices);
+}
+
+static if (__VERSION__ > 2069) @nogc unittest
+{
+    import std.algorithm: equal;
+    import std.range : iota;
+
+    import std.experimental.allocator.mallocator;
+
+    static immutable expected2 = [[0, 1], [1, 0]];
+    auto r = iota(2);
+    auto rp = makePermutations(Mallocator.instance, r.length);
+    assert(rp.indexedRoR(r).equal!equal(expected2));
+    dispose(Mallocator.instance, rp);
+}
+
+/// ditto
+struct Permutations
+{
+    private uint[] indices, state;
+    private bool _empty;
+
+    this(uint[] state, uint[] indices) @safe pure nothrow @nogc
+    in
+    {
+        assert(state.length + 1 == indices.length);
+    }
+    body
+    {
+        // iota
+        foreach(uint i, ref index; indices)
+            index = i;
+        state[] = 0;
+
+        this.indices = indices;
+        this.state = state;
+
+        _empty = indices.length == 0;
+    }
+
+    ///
+    @property uint[] front() @safe pure nothrow @nogc
```

Comment:
`const(uint)[] ` and for other structs too.
Also the note about that this range would be changed by the struct after `popFront`, for other structs too